lib/matchSourceExpression.js

```diff
@@ -0,0 +1,180 @@
+var urlModule = require('url');
+
+var ipv4DigitRegExp = /([0-9]|1?[0-9][0-9]|2[0-4][0-9]|25[0-5])/;
+var ipv4RegExp = new RegExp(
+    '^' +
+    ipv4DigitRegExp.source + '\\.' +
+    ipv4DigitRegExp.source + '\\.' +
+    ipv4DigitRegExp.source + '\\.' +
+    ipv4DigitRegExp.source + '$'
+);
+
+var defaultPortByScheme = {
+    http: 80,
+    https: 443
+};
+
+function origin(url) {
+    if (/file:\/\//.test(url)) {
+        return 'file:';
+    }
+    var matchOrigin = url.match(/^([a-z][a-z0-9+.-]+):\/\/(\*|(?:\*\.)?[a-z0-9-]+(?:\.[a-z0-9-]+)*)(?::(\d+|\*))?(\/.*)?$/i);
```

Comment:
These regexes quickly become unwieldy and hard to test. Could we pull this in from a well tested module?